interfaces/builtin/wayland.go

```diff
@@ -19,32 +19,160 @@
+	"github.com/snapcore/snapd/snap"
+)
+
 const waylandSummary = `allows access to compositors supporting wayland protocol`
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
+`
+
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is the display server, to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+/{dev,run}/shm/\#* mrw,
+
+/run/wayland-[0-9]* rw,
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# XWayland requires access to this
+/run/xwayland-shared-* rw,
+
+# Needed for mode setting via drmSetMaster() and drmDropMaster()
+capability sys_admin,
+
+# NOTE: this allows reading and inserting all input events
+/dev/input/* rw,
+
+# For using udev
+network netlink raw,
+/run/udev/data/c13:[0-9]* r,
+/run/udev/data/+input:input[0-9]* r,
+/run/udev/data/+platform:* r,
+`
+
+const waylandPermanentSlotSecComp = `
+# Description: Allow operating as a Wayland server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+# Needed by server upon client connect
+accept
+accept4
+shmctl
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedSlotAppArmor = `
+# Description: Permit clients to use Wayland
+unix (receive, send) type=seqpacket addr=none peer=(label=###PLUG_SECURITY_TAGS###),
 `
 
 const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+# Description: Permit clients to connect to compositors supporting the Wayland protocol
+unix (receive, send) type=seqpacket addr=none peer=(label=###SLOT_SECURITY_TAGS###),
 
-# Allow access to the wayland compsitor server socket
+# Allow access to the Wayland compositor server socket
+owner /run/wayland-[0-9]* rw,
 owner /run/user/*/wayland-[0-9]* rw,
 
+# XWayland needs access to this
+/run/xwayland-shared-* rw,
+
 # Needed when using QT_QPA_PLATFORM=wayland-egl
 /etc/drirc r,
 `
 
+type waylandInterface struct{}
+
+func (iface *waylandInterface) Name() string {
+	return "wayland"
+}
+
+func (iface *waylandInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              waylandSummary,
+		ImplicitOnClassic:    true,
+		BaseDeclarationSlots: waylandBaseDeclarationSlots,
+	}
+}
+
+func (iface *waylandInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	old := "###SLOT_SECURITY_TAGS###"
+	var new string
+	if release.OnClassic {
+		// If we're running on classic Wayland will be part
+		// of the OS snap and will run unconfined.
+		new = "unconfined"
+	} else {
+		new = slotAppLabelExpr(slot)
+	}
+	snippet := strings.Replace(waylandConnectedPlugAppArmor, old, new, -1)
```

Comment:
This code will change if the unix rule goes away.

Reply:
Done